sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala

```diff
@@ -80,6 +80,14 @@ class CacheManager extends Logging {
     cachedData.isEmpty
   }
 
+  private def extractStatsOfPlanForCache(plan: LogicalPlan): Option[Statistics] = {
+    if (plan.stats.rowCount.isDefined) {
```

Comment:
why this condition? Even there is no row count, the `sizeInBytes` is still better than `defaultSizeInBytes`

Reply:
I have the same question before. Seems @CodingCat's claim is that the `sizeInBytes` from relation can cause OOM issue: https://github.com/apache/spark/pull/19864#discussion_r155287758.

Reply:
If we don't trust `sizeInBytes`, we should fix it in all places, not just here

Reply:
+1

Reply:
We could also collect the size as part of building the cache.

Reply:
the current logic is that only when we have run have enough stats for the interested table, we would put the value here, 

I agree that we should choose a better value for format like parquet where the actual in-memory size would be much larger than the sizeInBytes (i.e. on-disk size), my question is 

1. what's the expected value if we have a HadoopFsRelation which is in parquet format ?

2. do we want to do it in this PR?

Reply:
The expected value should be `rowCount * avgRowSize`. Without CBO, I think the file size is the best we can get, although it may not be correct.

That is to say, without CBO, parquet relation may have underestimated size and cause OOM, users need to turn on CBO to fix it. So the same thing should happen in table cache.

We can fix this by defining `sizeInBytes` as `file size * some factor`, but it should belong to another PR.

Reply:
change the code here and file a JIRA in https://issues.apache.org/jira/browse/SPARK-22790